packages/webdriverio/src/utils/index.ts

```diff
@@ -230,8 +233,224 @@ export function isStaleElementError (err: Error) {
+        /**
+         * lookup in shadow roots
+         */
+        ...shadowRoots.map(
+            (shadowRootNodeId) => browser.findElementFromShadowRoot(shadowRootNodeId, using, value)
+                .then(
+                    elementPromiseHandler<ElementReference>(handle, shadowRootManager, shadowRootNodeId),
+                    elementPromiseHandler<ElementReference>(handle, shadowRootManager, shadowRootNodeId)
+                )
+                .then(async (res) => {
+                    if (!res) {
+                        return
+                    }
+
+                    /**
+                     * if an element was found with given selector within a shadow root
+                     * and
+                     * if we are calling the command from an element scope
+                     * then
+                     * we need to check if the found element is nested within the scope
+                     */
+                    if ('elementId' in this) {
+                        return (await isWithinElementScope(res, this as WebdriverIO.Element, handle))
+                            ? res
+                            : undefined
+                    }
+
+                    return res
+                })
+        )
+    ])
+
+    const allElementsFiltered = deepElementResult.flat().filter(Boolean) as ElementReference[]
+    if (allElementsFiltered.length === 0) {
+        return new Error(`Couldn't find element with selector "${selector}"`)
+    }
+
+    return allElementsFiltered[0] as ElementReference
+}
+
+/**
+ * Parallel look up of a selector within multiple shadow roots
+ * @param this WebdriverIO Browser or Element instance
+ * @param selector selector to look up
+ * @param isMulti set to true if you call from `$$` command
+ * @returns a list of shadow root ids with their corresponding matches or undefined if not found
+ */
+export async function findDeepElements(
+    this: WebdriverIO.Browser | WebdriverIO.Element,
+    selector: Selector
+): Promise<ElementReference[]> {
+    const browser = getBrowserObject(this)
+    const shadowRootManager = getShadowRootManager(browser)
+    const handle = await browser.getWindowHandle()
+
+    const shadowRoots = shadowRootManager.getShadowRootsForContext(handle)
+    const { using, value } = findStrategy(selector as string, this.isW3C, this.isMobile)
+    /**
+     * look up selector within document and all shadow roots
+     */
+    const deepElementsResult = await Promise.all([
+        /**
+         * standard lookup in document
+         */
+        browser.findElements(using, value),
+        /**
+         * lookup in shadow roots
+         */
+        ...shadowRoots.map(
+            (shadowRootNodeId) => browser.findElementsFromShadowRoot(shadowRootNodeId, using, value)
+                .then(
+                    elementPromiseHandler<ElementReference[]>(handle, shadowRootManager, shadowRootNodeId),
```

Comment:
I am not sure if this would make it that much more efficient, especially since elements change, show up or disappear. It makes it easier to just query elements again rather than caching results and worrying about when to clear the cache etc.